data/src/main/java/org/cryptomator/data/cloud/pcloud/PCloudImpl.java

```diff
@@ -0,0 +1,375 @@
+			} catch (IOException e) {
+				Timber.tag("PCloudImpl").w(e, "Error while retrieving content from Cache, get from web request");
+				writeToData(file, data, encryptedTmpFile, cacheKey, progressAware);
+			}
+		} else {
+			writeToData(file, data, encryptedTmpFile, cacheKey, progressAware);
+		}
+
+		progressAware.onProgress(Progress.completed(DownloadState.download(file)));
+	}
+
+	private void writeToData(final PCloudFile file, //
+			final OutputStream data, //
+			final Optional<File> encryptedTmpFile, //
+			final Optional<String> cacheKey, //
+			final ProgressAware<DownloadState> progressAware) throws IOException, BackendException {
+		try {
+			FileLink fileLink = client().createFileLink(file.getPath(), DownloadOptions.DEFAULT).execute();
+
+			ProgressListener listener = (done, total) -> progressAware.onProgress( //
+					progress(DownloadState.download(file)) //
+							.between(0) //
+							.and(file.getSize().orElse(Long.MAX_VALUE)) //
+							.withValue(done));
+
+			DataSink sink = new DataSink() {
+				@Override
+				public void readAll(BufferedSource source) {
+					CopyStream.copyStreamToStream(source.inputStream(), data);
+				}
+			};
+
+			client().download(fileLink, sink, listener).execute();
+		} catch(ApiError ex) {
+			handleApiError(ex);
+		}
+
+		if (sharedPreferencesHandler.useLruCache() && encryptedTmpFile.isPresent() && cacheKey.isPresent()) {
+			try {
+				storeToLruCache(diskLruCache, cacheKey.get(), encryptedTmpFile.get());
+			} catch (IOException e) {
+				Timber.tag("PCloudImpl").e(e, "Failed to write downloaded file in LRU cache");
+			}
+		}
+
+	}
+
+	public void delete(PCloudNode node) throws IOException, BackendException {
+		try {
+			if (node instanceof PCloudFolder) {
+				client() //
+						.deleteFolder(node.getPath(), true).execute();
+			} else {
+				client() //
+						.deleteFile(node.getPath()).execute();
+			}
+		} catch(ApiError ex) {
+			handleApiError(ex);
+		}
+	}
+
+	public String currentAccount() throws IOException, BackendException {
+		try {
+			UserInfo currentAccount = client() //
+					.getUserInfo() //
+					.execute();
+			return currentAccount.email();
+		} catch(ApiError ex) {
+			handleApiError(ex);
+			throw new FatalBackendException(ex);
+		}
+	}
+
+	private boolean createLruCache(int cacheSize) {
+		if (diskLruCache == null) {
+			try {
+				diskLruCache = DiskLruCache.create(new LruFileCacheUtil(context).resolve(PCLOUD), cacheSize);
+			} catch (IOException e) {
+				Timber.tag("PCloudImpl").e(e, "Failed to setup LRU cache");
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	private void handleApiError(ApiError ex) throws BackendException {
+		handleApiError(ex, null);
+	}
+
+	private void handleApiError(ApiError ex, Set<Integer> errorCodes) throws BackendException {
+		handleApiError(ex, errorCodes, null);
+	}
+
+	private void handleApiError(ApiError ex, Set<Integer> errorCodes, String name) throws BackendException {
```

Comment:
Do we need the `name`?

Reply:
It's pretty useless actually. The question would be what we hand over to `CloudNodeAlreadyExistsException()`.